rust/pact_matching/src/lib.rs

```diff
@@ -1520,10 +1519,12 @@ pub fn generate_request(request: &Request, mode: &GeneratorTestMode, context: &H
   if !generators.is_empty() {
     debug!("Applying header generators...");
     apply_generators(mode, &generators, &mut |key, generator| {
-      if let Some(ref mut headers) = request.headers {
-        if headers.contains_key(key) {
-          if let Ok(v) = generator.generate_value(&headers.get(key).unwrap().clone(), context, &DefaultVariantMatcher.boxed()) {
-            headers.insert(key.clone(), v);
+      if let Some(header) = key.first_field() {
```

Comment:
Is this an additional check that was needed?  I notice this is the first one off pattern.

Reply:
Yes. This has to do with how Header names and Query parameters are handled by the DocPaths and code surrounding them.

Previously, those strings were parsed by `calc_path_weight` at some point, which resulted in the token list `[Root, Field("HEADER")]` for example. But the original string `"HEADER"` was still passed around, including here as the variable `key`.

In this new way of doing things where the parsed tokens are preserved in `DocPath` structures (and `key` is now a `DocPath`), there needs to be a way to access the plain header name string, or query parameter string. I thought that it would be most robust to extract the string from the first Field token (which is what `DocPath::first_field()` does), rather than using the original string, currently stored as the field `expr` on `DocPath`. I thought this because I didn't want to depend on the current way that `DocPath`s are constructed from the plain strings, where `expr` just so happens to be the original plain string (without any Root token (`$`) or field separators (`.`)).